// @angular/language-service/ivy/compiler_factory.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/language-service/ivy/compiler_factory", ["require", "exports", "@angular/compiler-cli/src/ngtsc/core", "@angular/compiler-cli/src/ngtsc/incremental"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.CompilerFactory = void 0;
    var core_1 = require("@angular/compiler-cli/src/ngtsc/core");
    var incremental_1 = require("@angular/compiler-cli/src/ngtsc/incremental");
    /**
     * Manages the `NgCompiler` instance which backs the language service, updating or replacing it as
     * needed to produce an up-to-date understanding of the current program.
     *
     * TODO(alxhub): currently the options used for the compiler are specified at `CompilerFactory`
     * construction, and are not changable. In a real project, users can update `tsconfig.json`. We need
     * to properly handle a change in the compiler options, either by having an API to update the
     * `CompilerFactory` to use new options, or by replacing it entirely.
     */
    var CompilerFactory = /** @class */ (function () {
        function CompilerFactory(adapter, programStrategy, options) {
            this.adapter = adapter;
            this.programStrategy = programStrategy;
            this.options = options;
            this.incrementalStrategy = new incremental_1.TrackedIncrementalBuildStrategy();
            this.compiler = null;
            this.lastKnownProgram = null;
        }
        CompilerFactory.prototype.getOrCreate = function () {
            var _a;
            var program = this.programStrategy.getProgram();
            var modifiedResourceFiles = (_a = this.adapter.getModifiedResourceFiles()) !== null && _a !== void 0 ? _a : new Set();
            if (this.compiler !== null && program === this.lastKnownProgram) {
                if (modifiedResourceFiles.size > 0) {
                    // Only resource files have changed since the last NgCompiler was created.
                    var ticket_1 = core_1.resourceChangeTicket(this.compiler, modifiedResourceFiles);
                    this.compiler = core_1.NgCompiler.fromTicket(ticket_1, this.adapter);
                }
                else {
                    // The previous NgCompiler is being reused, but we still want to reset its performance
                    // tracker to capture only the operations that are needed to service the current request.
                    this.compiler.perfRecorder.reset();
                }
                return this.compiler;
            }
            var ticket;
            if (this.compiler === null || this.lastKnownProgram === null) {
                ticket = core_1.freshCompilationTicket(program, this.options, this.incrementalStrategy, this.programStrategy, 
                /* perfRecorder */ null, true, true);
            }
            else {
                ticket = core_1.incrementalFromCompilerTicket(this.compiler, program, this.incrementalStrategy, this.programStrategy, modifiedResourceFiles, /* perfRecorder */ null);
            }
            this.compiler = core_1.NgCompiler.fromTicket(ticket, this.adapter);
            this.lastKnownProgram = program;
            return this.compiler;
        };
        CompilerFactory.prototype.registerLastKnownProgram = function () {
            this.lastKnownProgram = this.programStrategy.getProgram();
        };
        return CompilerFactory;
    }());
    exports.CompilerFactory = CompilerFactory;
});
